File: iOS/DateMate/UI/Auth/SignIn/SignInViewModel.swift
//
//  SignInViewModel.swift
//  DateMate
//
//  Created by 홍희표 on 2021/11/28.
//

import Foundation
import FirebaseAuth
import Combine

class SignInViewModel: ObservableObject {
    private let userRepository: UserRepository
    
    @Published var uiState = SignInUiState()
    
    private var cancellables = Set<AnyCancellable>()

    private func isEmailValid(_ email: String) -> Bool {
        return email.range(of: SignInViewModel.EMAIL_VALIDATION_REGEX, options: .regularExpression) != nil
    }
    
    private func validateEmail(_ value: String) -> String? {
        if value.isEmpty || isEmailValid(value) {
            return nil
        } else {
            return "Invalid email: \(value)"
        }
    }

    private func isPasswordValid(_ password: String) -> Bool {
        return password.count > 3
    }

    private func validatePassword(_ value: String) -> String? {
        if value.isEmpty || isPasswordValid(value) {
            return nil
        } else {
            return "Invalid password"
        }
    }
    
    private func fetchUserProfile(_ userId: String) {
        userRepository.fetchUserProfile(userId: userId)
            .receive(on: DispatchQueue.main)
            .sink { resource in
                switch resource.state {
                case .Success:
                    let userCache = UserCache(id: userId)
                    
                    if resource.data != nil {
                        self.userRepository.storeUserProfile(userCache)
                        self.uiState.isLoading = false
                        self.uiState.message = nil
                    }
                case .Error:
                    self.uiState.isLoading = false
                    self.uiState.message = "로그인 성공, 프로필 로드 실패. 설정 화면으로 이동합니다."
                case .Loading:
                    self.uiState.isLoading = true
                    self.uiState.message = nil
                }
            }
            .store(in: &cancellables)
    }
    
    func onEmailChanged(_ value: String) {
        uiState.email = value
        uiState.emailError = validateEmail(value)
    }
    
    func onPasswordChanged(_ value: String) {
        uiState.password = value
        uiState.passwordError = validatePassword(value)
    }
    
    func signIn(_ email: String, _ password: String) {
        userRepository.getSignInResultStream(email: email, password: password)
            .receive(on: DispatchQueue.main)
            .sink { result in
                switch result.state {
                case .Loading:
                    self.uiState.isLoading = true
                    self.uiState.message = nil
                case .Success:
                    guard let firebaseUser = result.data else {
                        self.uiState.isLoading = false
                        self.uiState.message = "로그인 성공 후 사용자 정보를 찾을 수 없습니다."
                        return
                    }
                    self.fetchUserProfile(firebaseUser.uid)
                case .Error:
                    self.uiState.isLoading = false
                    self.uiState.message = result.message ?? "알 수 없는 로그인 오류가 발생했습니다."
                }
            }
            .store(in: &cancellables)
    }
    
    init(_ userRepository: UserRepository) {
        self.userRepository = userRepository
        
        // 이미 로그인된 사용자 체크 및 프로필 fetch
        userRepository.remoteUserStatePublisher
            .compactMap { $0 }
            .flatMap(maxPublishers: .max(1)) { user in
                userRepository.fetchUserProfile(userId: user.uid)
            }
            .sink { resource in
                switch resource.state {
                case .Success:
                    self.uiState.isAlreadySignIn = resource.data! == nil
                case .Error:
                    self.uiState.isAlreadySignIn = false
                    self.uiState.message = resource.message
                case .Loading: break
                }
            }
            .store(in: &cancellables)
    }
    
    private static let EMAIL_VALIDATION_REGEX = #"^(.+)@(.+)$"#
}

File: Android/app/src/main/java/com/hhp227/datemate/ui/auth/signin/SignInViewModel.kt
package com.hhp227.datemate.ui.auth.signin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.datemate.common.Resource
import com.hhp227.datemate.data.model.UserCache
import com.hhp227.datemate.data.repository.ProfileRepository
import com.hhp227.datemate.data.repository.UserRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.regex.Pattern

class SignInViewModel internal constructor(
    private val userRepository: UserRepository,
    private val profileRepository: ProfileRepository
) : ViewModel() {
    private val _uiState = MutableStateFlow(SignInUiState())
    val uiState: StateFlow<SignInUiState> = _uiState

    private fun isEmailValid(email: String): Boolean {
        return Pattern.matches(EMAIL_VALIDATION_REGEX, email)
    }

    private fun validateEmail(value: String): String? {
        return if (value.isEmpty() || isEmailValid(value)) null else "Invalid email: $value"
    }

    private fun isPasswordValid(password: String): Boolean {
        return password.length > 3
    }

    private fun validatePassword(value: String): String? {
        return if (value.isEmpty() || isPasswordValid(value)) null else "Invalid password"
    }

    private suspend fun fetchUserProfile(userId: String) {
        val profile = profileRepository.getProfile(userId)
        val userCache = UserCache(userId)

        if (profile != null) {
            userRepository.storeUserProfile(userCache)
            _uiState.update {
                it.copy(
                    isLoading = false,
                    message = null
                )
            }
        }
    }

    fun onEmailChanged(value: String) {
        _uiState.update {
            it.copy(
                email = value,
                emailError = validateEmail(value)
            )
        }
    }

    fun onPasswordChanged(value: String) {
        _uiState.update {
            it.copy(
                password = value,
                passwordError = validatePassword(value)
            )
        }
    }

    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            userRepository.getSignInResultStream(email, password)
                .collectLatest { result ->
                    when (result) {
                        is Resource.Error<*> -> {
                            _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    message = result.message ?: "알 수 없는 로그인 오류가 발생했습니다."
                                )
                            }
                        }
                        is Resource.Loading<*> -> _uiState.update { it.copy(isLoading = true, message = null) }
                        is Resource.Success<*> -> {
                            val firebaseUser = result.data

                            if (firebaseUser == null) {
                                _uiState.update { it.copy(isLoading = false, message = "로그인 성공 후 사용자 정보를 찾을 수 없습니다.") }
                                return@collectLatest
                            }
                            fetchUserProfile(firebaseUser.uid)
                        }
                    }
                }
        }
    }

    init {
        viewModelScope.launch {
            userRepository.remoteUserStateFlow
                .filterNotNull()
                .map { profileRepository.getProfile(it.uid) }
                .collectLatest { profile ->
                    _uiState.update { it.copy(isAlreadySignIn = profile == null) }
                }
        }
    }

    companion object {
        private const val EMAIL_VALIDATION_REGEX = "^(.+)@(.+)\$"
    }
}